Enable/Disable Propagation for Compiler Flags & Variants

RikkiButler20 opened the pull request:
Fixes #29653.

Allow for inheritance to be enabled or disabled for both variants and compiler flags.

- [x] Add new (spec language) operators
- [x] Apply operators to variants and compiler flags
- [x] Resolve conflicts
- [x] Add tests
- [x] Add documentation

RikkiButler20 committed:
flags/variants: Add ++/~~/== syntax for propagation to dependencies

Currently, compiler flags and variants are inconsistent: compiler flags set for a
package are inherited by its dependencies, while variants are not. We should have these
be consistent by allowing for inheritance to be enabled or disabled for both variants
and compiler flags.

- [x] Make new (spec language) operators
- [x] Apply operators to variants and compiler flags
- [x] Conflicts currently result in an unsatisfiable spec
      (i.e., you can't propagate two conflicting values)

What I propose is using two of the currently used sigils to symbolized that the variant
or compiler flag will be inherited:

Example syntax:
- `package ++variant`
      enabled variant that will be propagated to dependencies
- `package +variant`
      enabled variant that will NOT be propagated to dependencies
- `package ~~variant`
      disabled variant that will be propagated to dependencies
- `package ~variant`
      disabled variant that will NOT be propagated to dependencies
- `package cflags==True`
      `cflags` will be propagated to dependencies
- `package cflags=True`
      `cflags` will NOT be propagated to dependencies

Syntax for string-valued variants is similar to compiler flags.

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -943,10 +965,12 @@ compiler_weight(Package, 100)
 % propagate flags when compilers match
 inherit_flags(Package, Dependency)
- :- depends_on(Package, Dependency),
+ :- path(Package, Dependency),
```
This change should be unnecessary, because all packages have compiler flags (where it was necessary for variants to use `path` because the intermediate package might not have it).

RikkiButler20 replied:
When I use `depends_on` the compiler flag is only passed to the direct descendants of the root package.

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -943,10 +965,12 @@ compiler_weight(Package, 100)
     node_compiler(Package, Compiler),
     node_compiler(Dependency, Compiler),
-    compiler(Compiler), flag_type(FlagType).
+    compiler(Compiler), flag_type(FlagType),
```
```suggestion
```

These facts can be inferred from `node_compiler(Package, Compiler)` and `node_flag_propagate(Package, FlagType)`, there's no need to duplicate them here.

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -480,6 +484,23 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
+  variant_possible_value(Descendant, Variant, Value).
+
+
+:- variant_propagate(Package, Variant),
+   variant_propagate(Descendant, Variant),
+   path(Package, Descendant),
+   build(Package),
+   error("Cannot have variant propagation on parent and descendant"). % Only if the same variant
```
This constraint needs to be rewritten to infer an error, rather than failing the solve. I think the issue is that you did it the old way before the error handling PR changed things, and it needs to change to match other constraints.

I think the next constraint is a good model for this, something like 
```suggestion
error(2, "{0} and dependency {1} cannot both propagate variant '{2}'", Package, Descendant, Variant) :- 
  variant_propagate(Package, Variant),
  variant_propagate(Descendant, Variant),
  path(Package, Descendant),
  build(Package).
```

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1253,6 +1253,8 @@ class Head(object):
             node_compiler = fn.node_compiler_set
             node_compiler_version = fn.node_compiler_version_set
             node_flag = fn.node_flag_set
+            node_flag_propagate = fn.node_flag_propagate
+            variant_propagate = fn.variant_propagate
```
These probably need to be added to the class `Body` as well

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1962,19 +1969,20 @@ def error(self, priority, msg, *args):
         raise UnsatisfiableSpecError(msg)
 
-    def variant_value(self, pkg, name, value):
+    def variant_value(self, pkg, name, value, propagate=False):
```
I don't think this `propagate` argument is actually doing anything... `variant_value` facts in the concretizer don't have a fourth field, so it will never be `True`. And it's only passed to `patches` and `dev_path` variants, which would never be propagated since they're generated by Spack internally.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -52,8 +53,10 @@
+  options      = { @version-list | ++variant | +variant |
+                   --variant | -variant | ~~variant | ~variant |
+                   %compiler | arch=architecture |
+                   [ flag ]==value | [ flag ]=value}
```
Side note: This grammar is incorrect (prior to this PR) because it doesn't account for variants specified by `name=value`.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -151,6 +155,10 @@
 identifier_re = r'\w[\w-]*'
 
+
+# TODO: Should we add another color for the
```
I don't think that's worth pursuing

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -716,9 +724,10 @@ def canonical(self):
+    def __init__(self, spec, propagate=False):
         super(FlagMap, self).__init__()
         self.spec = spec
+        self.propagate = propagate
```
This looks like it sets propagate for all compiler flags, rather than individually. What about `spack spec tcl cflags==-g cxxflags=-O3`? I think we need a test to confirm that propagates the appropriate one but not the other.

RikkiButler20 replied:
You're right. Neither compiler flag propagated. Thanks for catching that.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -1401,15 +1422,16 @@ def _add_flag(self, name, value):
         elif name in valid_flags:
             assert(self.compiler_flags is not None)
             self.compiler_flags[name] = spack.compiler.tokenize_flags(value)
+            self.compiler_flags.propagate = propagate
```
```suggestion
            self.compiler_flags.propagate[name] = propagate
```
The `FlagMap` class would also need to be updated to handle this.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -5245,10 +5272,18 @@ def spec(self, name):
 
+            elif self.accept(D_ON):
+                name = self.variant()
+                spec.variants[name] = vt.BoolValuedVariant(name, True, True)
```
```suggestion
                spec.variants[name] = vt.BoolValuedVariant(name, True, propagate=True)
```
Just a little easier to read this way

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -5245,10 +5272,18 @@ def spec(self, name):
 
+            elif self.accept(D_OFF):
+                name = self.variant()
+                spec.variants[name] = vt.BoolValuedVariant(name, False, True)
```
```suggestion
                spec.variants[name] = vt.BoolValuedVariant(name, False, propagate=True)
```

becker33 commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -328,6 +335,7 @@ def test_satisfies_single_valued_variant(self):
         # Assert that an autospec generated from a literal
         # gives the right result for a single valued variant
         assert 'foobar=bar' in a
+        assert 'foobar==bar' in a
```
Does this just mean that `foobar=bar` satisfies `foobar==bar`? This seems like an odd place to be testing that, but if that's all it's testing it might be ok here.

becker33 reviewed (changes requested):
This mostly LGTM, a couple of concerns about potentially vestigial code and some legibility suggestions.

becker33 commented on lib/spack/spack/test/concretize.py:
```diff
@@ -417,6 +417,24 @@ def test_concretize_two_virtuals_with_dual_provider(self):
         Spec('hypre ^openblas-with-lapack').concretize()
 
+    def test_concretize_propagate_disabled_variant(self):
+        """Test a package with a disabled propagated variant."""
+        Spec('hypre~~shared ^openblas').concretize()
+
+    def test_concretize_propagated_disabled_variant_passed_to_dependent(self):
```
```suggestion
```

I think the second test is sufficiently simple that it's not worth separating the two.

tgamblin commented:
@spackbot fix styl

tgamblin commented:
@spackbot fix style

scottwittenburg commented:
@spackbot fix style

RikkiButler20 commented:
@spack-bot fix style

RikkiButler20 commented:
@spackbot fix style

becker33 commented on lib/spack/spack/build_environment.py:
```diff
@@ -343,10 +343,11 @@ def set_compiler_environment_variables(pkg, env):
         if env_flags[flag]:
             # implicit variables
-            env.set(flag.upper(), " ".join(f for f in env_flags[flag]))
+            env.set(flag.upper(), " ".join(f[0] for f in env_flags[flag]))
```
I realized this implementation is going to break custom flag handlers.

We either need to be able to iterate over the flags and their propagate values separately or we need to only pass the flags (not the propagate values) to the flag handler.

We could switch from `[(, ), (, )]` to two ordered lists, or we could call the flag handler differently. And then we can remove these changes.

I'm partial to the option where we store two lists, because it allows package writers to access the flags in a more intuitive way. Basically `compiler_flags[name]` would give just the flags, and we add another data structure `compiler_flags._propagation[name]` gives the propagation values. Then we can define `compiler_flags.with_propagation()` to return the tuples the way you've constructed them, for places we need that information.

We can overwrite whatever methods we need to be able to extend this interface to other ways we interact with the `FlagMap` object.

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1290,6 +1271,9 @@ class Body(object):
                 clauses.append(f.variant_value(spec.name, vname, value))
 
+                if variant.propagate:
+                    clauses.append(f.variant_propagate(spec.name, vname))
```
Eventually we should make the propagation be per-flag, but that's future work in another PR. Then we can enable

```
spack install foo cflags=-g cflags==-O3
```
to have the root build with `-g -O3` and the dependencies just with `-O3`.

RikkiButler20 replied:
@becker33 So do you think `node_flag_propagate` should include the flag_value in the future?

becker33 replied:
Yes, exactly

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1315,7 +1299,9 @@ class Body(object):
             for flag in flags:
-                clauses.append(f.node_flag(spec.name, flag_type, flag))
+                clauses.append(f.node_flag(spec.name, flag_type, flag[0]))
+                if flag[1] is True:
+                    clauses.append(f.node_flag_propagate(spec.name, flag_type))
```
We shouldn't regenerate this for each individual flag
```suggestion
            if flags and flags[0][1] is True:
                clauses.append(f.node_flag_propagate(spec.name, flag_type))
```

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2032,14 +2036,16 @@ def reorder_flags(self):
+                spec.compiler_flags[flag_type],
+                flags,
+            )
+            spec.compiler_flags.update({flag_type: flags})
```
This probably needs an update for when compiler flags also came from the compiler definition, but that can wait for a follow-on, it's a preexisting bug.

becker33 commented on lib/spack/spack/spec.py:
```diff
@@ -35,6 +35,8 @@
    adding +debug to the openmpi spec when you depend on it.  If you do
    NOT want the debug option to be enabled, then replace this with -debug.
+   If you would like for the variant to be propagated through all your
+   package's dependencies "++" for enabling and "--" or "~~" for disabling.
```
```suggestion
   package's dependencies use "++" for enabling and "--" or "~~" for disabling.
```

becker33 reviewed (changes requested):
I think the internal representation change I've requested here would also allow you to eliminate most of the weirder changes in tests that were designed to merely preserve previous behavior

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot fix style

adamjstewart commented:
@becker33 pointed me to this, so I'm seeing this for the first time. Trying to catch up on the discussion and motivation for the design features introduced in this PR.

Instead of introducing a new spec syntax (`++`, `~~`), why not just propagate by default and allow users to override? For example:

```console
$ spack install netcdf-c +mpi  # build netcdf-c and hdf5 with +mpi
$ spack install netcdf-c -mpi  # build netcdf-c and hdf5 with -mpi
$ spack install netcdf-c +mpi ^hdf5 -mpi  # build netcdf-c with +mpi and hdf5 with -mpi
```
(netcdf-c and hdf5 aren't the best example because we have variant forwarding hard-coded into the package, but imagine we removed this)

This is how compilers already work. This already works in `packages.yaml`:
```yaml
packages:
  all:
    variants: +mpi
  hdf5:
    variants: ~mpi
```

tgamblin commented:
@adamjstewart: there was a pretty long discussion about this on Slack a while back but unfortunately it was eaten.

We have a lot of requests for:
- The ability to propagate variants
- The ability to NOT propagate compilers

Right now the default for variants is not to propagate, and you can't actually propagate.  The default for compilers is to propagate, and you can't choose not to.  We want a solution that is consistent and easy to understand that also gives users control.  Also:

1. In general users usually *don't* want compiler flags to propagate, at least not to all their dependencies.  Frequently we see people *just* wanting to experiment with flags on the root and not with dependencies (if only for build time, which is a big deal).
2. Variant propagation by default is similarly often wrong.  e.g. many MPI codes require `hdf5~mpi`; many codes that use `openmp` want to use a non-threaded BLAS implementation, etc.

To make propagation and non-propagation cases simple and understandable, we have to break compilers or variants -- we went with breaking compilers because we were seeing that more often than not people don't want propagation when they specify flags.

So:

1. These will not propagate:
    * `foo +bar`
    * `foo ~bar`
    * `foo -bar`
    * `foo bar=True`
    * `foo bar=False`
    * `foo cflags="-g -O3"`
2. These will propagate:
    * `foo ++bar`
    * `foo ~~bar`
    * `foo --bar`
    * `foo bar==True`
    * `foo bar==False`
    * `foo cflags=="-g -O3"`

If we made everything propagate as in your example, the user doesn't really have control over the subgraph.  For example in your HDF5 example:

```console
$ spack install netcdf-c +mpi  # build netcdf-c and hdf5 with +mpi
$ spack install netcdf-c -mpi  # build netcdf-c and hdf5 with -mpi
$ spack install netcdf-c +mpi ^hdf5 -mpi  # build netcdf-c with +mpi and hdf5 with -mpi
```

The last comment is actually not correct.  If we propagate by default there, then `hdf5` *and all of its dependencies* will build with `-mpi`.  With the syntax we're proposing here, it's much clearer what should happen:

```console
$ spack install netcdf-c ++mpi ^hdf5 -mpi  # build netcdf-c with +mpi and *just* hdf5 with -mpi
```

To express that with default propagation you'd need to write out `^hdf5 -mpi` and `^dep +mpi` for *all* of hdf5's dependencies.  The only backward-incompatible change we're introducing here is that we're changing the default propagation semantics for compiler flags, which IMO makes sense given user demand and our desire to be consistent with variants.

RikkiButler20 commented:
@spackbot fix style

alalazo commented:
What is the behavior we have in this PR for conflicting variant propagation coming to a node through different edges? Are we erroring out in that case for the time being?

RikkiButler20 commented:
@alalazo The same variant cannot be propagated with different values for now, so this will not be an issue. A final solution will. be in a subsequent PR.

tgamblin commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot re-run pipeline

RikkiButler20 commented:
@spackbot re-run pipeline

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot fix style

RikkiButler20 commented:
@spackbot re-run pipeline

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -999,10 +999,13 @@ More formally, a spec consists of the following pieces:
   (``gcc`` or ``gcc@4.7.3``)
 * ``+`` or ``-`` or ``~`` Optional variant specifiers (``+debug``,
   ``-qt``, or ``~qt``) for boolean variants
+  Use ``++`` or ``--`` or ``~~`` to propagate variants through the dependecies.
```
Do you want this to appear with the previous bullet at https://spack--29761.org.readthedocs.build/en/29761/basic_usage.html#specs-dependencies or a new version?

If the same, then add a period (`.`) to line 1001; otherwise, start the line with an asterisk (`*`) and maybe add examples.

Also, fix the `dependencies` typo.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -999,10 +999,13 @@ More formally, a spec consists of the following pieces:
+  Use ``++`` or ``--`` or ``~~`` to propagate variants through the dependecies.
 * ``name=<value>`` Optional variant specifiers that are not restricted to
   boolean variants
+  Use ``name==<value>`` to propagate variant through the dependecies.
```
Same issues and suggestions as line 1002.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -999,10 +999,13 @@ More formally, a spec consists of the following pieces:
+  Use ``name==<value>`` to propagate variant through the dependecies.
 * ``name=<value>`` Optional compiler flag specifiers. Valid flag names are
   ``cflags``, ``cxxflags``, ``fflags``, ``cppflags``, ``ldflags``, and ``ldlibs``.
+  Use ``name==<value>`` to propagate compiler flags through the dependecies.
```
```suggestion
  Use ``name==<value>`` to propagate compiler flags through the dependencies.
```

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1226,17 +1229,27 @@ variants using the backwards compatibility syntax and uses only ``~``
 
+Spack allows variants to propagate their value to the package's
+dependency by using ``++``, ``--``, and ``~~`` for boolean variants,
+and ``name==value`` for non-boolean variants. For example:
```
Perhaps reword to "For example, for a `debug` variant:".

Though it might be worth adding examples for the `stackstart` variant (e.g., `startstart==4`) since we've used that value in the tutorial at https://spack-tutorial.readthedocs.io/en/latest/tutorial_packaging.html#adding-variants.

RikkiButler20 replied:
@tldahlgren Are you suggesting replacing the `debug` example with `stackstart` or adding a `stackstart` example? I think adding the `stackstart` example would be best so we can have examples with `++` and `==`.

tldahlgren replied:
I agree adding is better.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1226,17 +1229,27 @@ variants using the backwards compatibility syntax and uses only ``~``
-install both libdwarf and libelf with the ``-g`` flag injected into their
-compile line.
+flags into the compile line of the build. Additionally, compiler flags can
+be inherited by dependencies by using ``==``. Though, versions of spack prior to
+0.19.0 will use the ``=`` syntax.
```
Perhaps the reference to pre-0.19.0 versions could be added as a `note` for emphasis instead?

RikkiButler20 replied:
@tldahlgren I'm confused on what a reference to pre-0.19.0 would be. Do you mean a link, or something else?

tldahlgren replied:
Resolving since we discussed this off-line.

tldahlgren commented on lib/spack/spack/compiler.py:
```diff
@@ -311,11 +311,14 @@ def __init__(
-        for flag in spack.spec.FlagMap.valid_compiler_flags():
+        self.flags = spack.spec.FlagMap(self.spec)
+        for flag in self.flags.valid_compiler_flags():
+            # TODO: This should be a FlagMap not dict
```
Is this comment still relevant?

RikkiButler20 replied:
@tldahlgren The comment is in reference to `kwargs`. I can remove the comment, but this should be fixed at some point, even if it's not in this PR.

tldahlgren commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -72,7 +72,11 @@ def _to_dict(compiler):
     d["spec"] = str(compiler.spec)
     d["paths"] = dict((attr, getattr(compiler, attr, None)) for attr in _path_instance_vars)
-    d["flags"] = dict((fname, fvals) for fname, fvals in compiler.flags)
+    # TODO: This shouldnt be a dictionary, but a FlagMap
```
Is this comment still relevant?  If so, is there more work to be done here?

tldahlgren commented on lib/spack/spack/concretize.py:
```diff
@@ -564,7 +564,7 @@ def concretize_compiler_flags(self, spec):
                 raise
             return ret
-        for flag in compiler.flags:
+        for flag in compiler.flags.keys():
```
Is use of `.keys()` required here?

tldahlgren commented on lib/spack/spack/package_base.py:
```diff
@@ -2620,7 +2620,7 @@ def do_activate(self, view=None, with_dependencies=True, verbose=True):
 
         if verbose:
-            tty.debug(
+            tty.msg(
```
Should this be returned to a `debug` message?

tldahlgren commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2151,15 +2134,24 @@ def reorder_flags(self):
             compiler_flags = compilers[spec.compiler].flags
-            check_same_flags(spec.compiler_flags, compiler_flags)
-            spec.compiler_flags.update(compiler_flags)
+            for key in spec.compiler_flags.keys():
```
As mentioned above, is `.keys()` actually required here?

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -773,18 +796,37 @@ def constrain(self, other):
             elif k not in self:
                 self[k] = other[k]
                 changed = True
+            # Check that propagation
```
Did you want to complete this comment (i.e., `Check that propagation <something>`)?

tldahlgren commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2169,14 +2161,16 @@ def reorder_flags(self):
 
             # add flags from each source, lowest to highest precedence
-            flags = collections.defaultdict(lambda: [])
+            flags = []
```
Should there be an explanation for why this isn't a dictionary?

RikkiButler20 replied:
@tldahlgren I'm not sure why. What part needs to be explained? We're just checking the flag values, so which grab the entire dictionary?

tldahlgren replied:
Resolving since we discussed this offline too.

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -773,18 +796,37 @@ def constrain(self, other):
+            clone[name] = compiler_flag
         return clone
 
+    def add_flag(self, flag_type, value, propagation):
```
Perhaps add a docstring here to explain the method and arguments since this is a brand new method (e.g., see the docstring for `satisfies()` at ~line 399.

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -735,6 +740,13 @@ def canonical(self):
         return self.parent.dag_hash(), self.spec.dag_hash(), self.deptypes
 
 
+class CompilerFlag(str):
```
Suggest adding a description (docstring) of the class / it's purpose since it is new.

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -773,18 +796,37 @@ def constrain(self, other):
+        else:
+            self[flag_type].append(flag)
+
+    def yaml_entry(self, flag_type):
```
Suggest adding a description of the argument here as well.

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -4269,7 +4342,7 @@ def write_attribute(spec, attribute, color):
                             parent = ".".join(parts[:idx])
                             m = "Attempted to format attribute %s." % attribute
-                            m += "Spec.%s has no attribute %s" % (parent, part)
+                            m += "Spec %s has no attribute %s" % (parent, part)
```
Great catch!

tldahlgren reviewed (commented):
Mostly documentation-related changes (e.g., typos, docstrings) suggested.

becker33 commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1226,17 +1230,35 @@ variants using the backwards compatibility syntax and uses only ``~``
-install both libdwarf and libelf with the ``-g`` flag injected into their
-compile line.
+flags into the compile line of the build. Additionally, compiler flags can
+be inherited by dependencies by using ``==``. Though, versions of spack prior to
+0.19.0 will use the ``=`` syntax.
```
```suggestion
be inherited by dependencies by using ``==``. Note: versions of spack prior to
0.19.0 will propagate compiler flags using the ``=`` syntax.
```

RikkiButler20 replied:
@becker33 Tammy suggested I put this in a note using `.. note`, I just haven't don that yet. 😅

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -559,6 +565,22 @@ error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
+  not variant_set(Descendant, Variant),
+  variant_possible_value(Descendant, Variant, Value).
+
+error(2, "{0} and dependency {1} cannot both propagate variant '{2}'", Package, Descendant, Variant) :- 
+  variant_propagate(Package, Variant),
+  variant_propagate(Descendant, Variant),
+  path(Package, Descendant),
+  build(Package).
```
I'm not sure why this is necessary

RikkiButler20 replied:
@becker33 This is to prevent having the same variant propagate with different values in a spec. Like this:
`spack spec openssl ++docs ^berkeley-db ~~docs`

becker33 replied:
We definitely want to avoid multiple values propagating to the same node. What about

```
error(2, "{0} and {1} propagate incompatible values {2} and {3} for variant {4} to {5}", Package1, Package2, Value1, Value2, Variant, Descendant) :-
  variant_propagate(Package1, Variant),
  variant_propagate(Package2, Variant),
  variant_value(Package1, Variant, Value1),
  variant_value(Package2, Variant, Value2),
  value1 != value2,
  path(Package1, Descendent),
  path(Package2, Descendent),
  build(Descendent).
```

I'm concerned that the current versions prevents `foo ++shared ^bar~~shared ^baz` but it doesn't prevent two unrelated parents propagating to the same place.

For a test, maybe we should have a test that concretizes `dt-diamond ^dt-diamond-right ++shared ^dt-diamond-left ~~shared` and ensure that it throws a reasonable error.

Everything below here should probably be in a separate PR:

In the long run, we'll want to come back to this and enable propagation to be overridden for a sub-DAG. For example, if `foo -> bar -> baz -> corge`, we should be able to do `foo ++shared ^baz ~~shared` and result in

`foo+shared ^bar+shared ^baz~shared ^corge~shared`.

For that, we could have something like "inherit if there's a path and there's not an intermediate package that also propagates this variant in that path". But that's futureware.

RikkiButler20 replied:
@becker33 Thanks for catching this. I'll make some changes, but I'm not sure that we need to compare `value1` and `value2`. Shouldn't it fail regardless?

becker33 replied:
that should be `Value1 != Value2`, I messed up the capitalization. I think we want a check for that, otherwise something like 

`hdf5++shared ^mpich++shared` 

will fail.

It's not the end of the world if that fails, but it seems like it's nearly trivial to catch that case as one we don't mind.

RikkiButler20 replied:
@becker33 But wouldn't shared already propagate to `mpich`'s dependencies because of `hdf5++shared`?

becker33 commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -554,6 +574,14 @@ def test_constrain_variants(self):
         check_constrain("libelf+debug~foo", "libelf+debug", "libelf~foo")
         check_constrain("libelf+debug~foo", "libelf+debug", "libelf+debug~foo")
 
+        check_constrain("libelf++debug++foo", "libelf++debug", "libelf+foo")
```
Shouldn't `Spec('libelf++debug').constraint('libelf+foo') == Spec('libelf++debug+foo')?

alalazo commented:
Did anybody measure the performance impact of this PR? As a point measurement, with this PR:
```console
$ spack solve --timers trilinos
Time:
    setup:         9.5961
    load:          0.0227
    ground:        5.8712
    solve:         6.7784
Total: 22.4854
[ ... ]
```
while on `develop`:
```console
$ spack solve --timers trilinos
Time:
    setup:         9.4514
    load:          0.0220
    ground:        3.7320
    solve:         5.2694
Total: 18.6881
[ ... ]
```
This is a 20% overall slowdown, stemming from a ~60% slowdown in grounding and a ~25% slowdown in solving. I get the same kind of results, qualitatively, from a run on the specs in [radiuss.txt](https://github.com/spack/spack/files/9888085/radiuss.txt).

becker33 committed:
propagation: improve performance

This updates the propagation logic used in `concretize.lp` to avoid rules with `path()`
in the body and instead base propagation around `depends_on()`.

becker33 commented:
@alalazo I tried a minor refactor of the key propagation lines in the concretizer, and in my testing so far I'm seeing performance in line with develop. Let me know how it looks to you.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1255,11 +1255,15 @@ Compiler flags are specified using the same syntax as non-boolean variants,
 
+.. note::
+
+   versions of spack prior to 0.19.0 will propagate compiler flags using
```
Suggest capitalizing the first letter of the first word since this text is in a note.

```suggestion
   Versions of spack prior to 0.19.0 will propagate compiler flags using
```

tgamblin commented:
@spackbot fix style

tgamblin commented:
I've cleaned up the commits here and squashed into the original implementation + @becker33's performance fix.  With the fix, I see no additional overhead for this feature, other than that setup seems to take a little longer:

`develop`:
```console
(spackbook):spack> spack solve --timers trilinos
Time:
    setup:         5.9787
    load:          0.0143
    ground:        1.8304
    solve:         3.0472
Total: 10.9299
```

This PR:
```console
(spackbook):spack> spack solve --timers trilinos
Time:
    setup:         6.3022
    load:          0.0138
    ground:        1.9173
    solve:         3.0648
Total: 11.3539
```

Setup is going to get optimized away soon, so I'm fine with this as-is.

tgamblin commented:
RHEL8 issue here is fixed on develop; all PR-relevant tests are passing.

tgamblin commented:
@alalazo i think a follow on PR that doesn’t count propagated variants as non-default would address the biggest issue you list here.

haampie commented:
This PR (also with follow up PRs) slows down unit tests very significantly.

The slowest test

```
86.64s call     lib/spack/spack/test/concretize_preferences.py::TestConcretizePreferences::test_preferred_compilers
```
now takes

```
182.03s call     lib/spack/spack/test/concretize_preferences.py::TestConcretizePreferences::test_preferred_compilers
```

This individual test now runs slower than all 3500 other tests combined when run with xdist, which is quite annoying for development (2 mins of waiting -> 3+ minutes)

tgamblin commented:
If we can’t make a general fix for this, we can separate the propagation logic introduced here from `concretize.lp` and only include it if we see propagation used. Then we would at least only pay when using this feature.
